materialize-snowflake/snowflake.go

```diff
@@ -717,113 +717,105 @@ func (d *transactor) Acknowledge(ctx context.Context) (*pf.ConnectorState, error
-				// we might have processed all files of this pipe previously
-				if len(pipe.files) == 0 {
-					continue
-				}
-
-				// We first try to check the status of pipes using the REST API's insertReport
-				// The REST API does not wake up the warehouse, hence our preference
-				// however this API is not the most ergonomic and sometimes does not yield
-				// results as expected
-				report, err := d.pipeClient.InsertReport(pipeName)
-				if err != nil {
-					return nil, fmt.Errorf("snowpipe: insertReports: %w", err)
-				}
-
-				// If the files have already been loaded, when we submit a request to
-				// load those files again, our request will not show up in reports.
-				// Moreover, insertReport only retains events for 10 minutes.
-				// One way to find out whether the files were successfully loaded is to check
-				// the COPY_HISTORY to make sure they are there. The COPY_HISTORY is much more
-				// reliable. If they are not there, then something is wrong.
-				if len(report.Files) == 0 {
-					// We try `maxTries` times since it may take some time for the REST API
-					// to reflect the new pipe requests
-					if tries < maxTries-1 {
-						time.Sleep(retryDelay)
-						continue
-					}
+	// If we see no results from the REST API for 5 minutes, then we fallback to asking the `COPY_HISTORY` table
+	var retryDelay = 500 * time.Millisecond
+	var maxTryTime = time.Now().Add(5 * time.Minute)
+
+	for len(pipes) > 0 {
+		for pipeName, pipe := range pipes {
+			// if the pipe has no files to begin with, just skip it
+			// we might have processed all files of this pipe previously
+			if len(pipe.files) == 0 {
+				delete(pipes, pipeName)
+				continue
+			}
 
-					log.WithFields(log.Fields{
-						"tries": tries,
-						"pipe":  pipeName,
-					}).Info("snowpipe: no files in report, fetching copy history from warehouse")
+			// We first try to check the status of pipes using the REST API's insertReport
+			// The REST API does not wake up the warehouse, hence our preference
+			// however this API is limited to 10,000 results and the last 10 minutes only
+			report, err := d.pipeClient.InsertReport(pipeName)
+			if err != nil {
+				return nil, fmt.Errorf("snowpipe: insertReports: %w", err)
+			}
 
-					var fileNames = make([]string, len(pipe.files))
-					for i, f := range pipe.files {
-						fileNames[i] = f.Path
-					}
+			// If the files have already been loaded, when we submit a request to
+			// load those files again, our request will not show up in reports.
+			// Moreover, insertReport only retains events for 10 minutes.
+			// One way to find out whether files were successfully loaded is to check
+			// the COPY_HISTORY table to make sure they are there. The COPY_HISTORY is much more
+			// reliable. If they are not there, then something is wrong.
+			if len(report.Files) == 0 {
+				if time.Now().Before(maxTryTime) {
+					time.Sleep(retryDelay)
+					continue
+				}
```

Comment:
Check for time elapsed to decide whether it's time to check `COPY_HISTORY` table in case of no results from API